File: src/RepositoryModules.CS/ModeConfigForm1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using Storex;

namespace RepositoryModules
{
    partial class ModeConfigForm1
    {
        static readonly IReadOnlyList<KeyValuePair<string, SecondaryLabelBehavior>> SecondaryLabelBehaviorViewModels = new[]
        {
            new KeyValuePair<string, SecondaryLabelBehavior>("許可", SecondaryLabelBehavior.Default),
            new KeyValuePair<string, SecondaryLabelBehavior>("登録時に警告を表示", SecondaryLabelBehavior.Warnning),
            new KeyValuePair<string, SecondaryLabelBehavior>("拒否", SecondaryLabelBehavior.Deny)
        };

        public ModeConfigForm1()
        {
            InitializeComponent();
        }

        void OkButton_Click(object sender, EventArgs e)
        {
            if (string.IsNullOrEmpty(PrimaryPrefixKeyTextBox.Text.TrimEnd()))
            {
                ErrorProvider.SetError(PrimaryPrefixKeyTextBox, "入力してください");
            }
            else
            {
                ErrorProvider.SetError(PrimaryPrefixKeyTextBox, null);
            }

            DialogResult = DialogResult.OK;
        }

        public static void Configure(IMode Mode)
        {
            FixedLengthSpec primaryLabelSpec = Mode.TryExtractProperty(FixedLengthSpec.PropertyKeyForPrimary, out primaryLabelSpec) ?
                primaryLabelSpec :
                new FixedLengthSpec();

            SecondaryLabelCriteria secondaryLabelCriteria = Mode.TryExtractProperty(SecondaryLabelCriteria.PropertyKey, out secondaryLabelCriteria) ?
                secondaryLabelCriteria :
                new SecondaryLabelCriteria();

            using (var form = new ModeConfigForm1())
            {
                form.PrimaryPrefixKeyTextBox.Text = primaryLabelSpec.PrefixKey;
                form.PrimaryItemStartUpDown.Value = primaryLabelSpec.ItemNumberStartIndex;
                form.PrimaryItemLengthUpDown.Value = primaryLabelSpec.ItemNumberLength;
                form.PrimarySerialStartUpDown.Value = primaryLabelSpec.SerialNumberStartIndex;
                form.PrimarySerialLengthUpDown.Value = primaryLabelSpec.SerialNumberLength;

                form.AcceptC3LabelCheckBox.Checked = (secondaryLabelCriteria.AcceptableTypes & SecondaryLabelTypes.C3Label) == SecondaryLabelTypes.C3Label;
                form.AcceptSingleSymbolLabelCeckBox.Checked = (secondaryLabelCriteria.AcceptableTypes & SecondaryLabelTypes.SingleSymbol) == SecondaryLabelTypes.SingleSymbol;

                form.EqualsToPrimaryComboBox.DataSource = SecondaryLabelBehaviorViewModels.ToList();
                form.SpecifiedByConditionComboBox.DataSource = SecondaryLabelBehaviorViewModels.ToList();
                form.OtherNotSinglLabelComboBox.DataSource = SecondaryLabelBehaviorViewModels.ToList();

                form.NoLabelBehaviorComboBox.DataSource = new[]
                {
                    new KeyValuePair<string, SecondaryNoLabelBehavior>("許可", SecondaryNoLabelBehavior.Default),
                    new KeyValuePair<string, SecondaryNoLabelBehavior>("登録時に警告を表示", SecondaryNoLabelBehavior.Warnning),
                    new KeyValuePair<string, SecondaryNoLabelBehavior>("登録時にタグで品名確認(一致しないとき警告)", SecondaryNoLabelBehavior.TagRequest),
                    new KeyValuePair<string, SecondaryNoLabelBehavior>("登録時にタグで品名確認(一致しないとき拒否)", SecondaryNoLabelBehavior.TagRequired),
                    new KeyValuePair<string, SecondaryNoLabelBehavior>("拒否", SecondaryNoLabelBehavior.Deny)
                };

                form.NoLabelBehaviorComboBox.DisplayMember
                    = form.EqualsToPrimaryComboBox.DisplayMember
                    = form.SpecifiedByConditionComboBox.DisplayMember
                    = form.OtherNotSinglLabelComboBox.DisplayMember = "Key";

                form.NoLabelBehaviorComboBox.ValueMember
                    = form.EqualsToPrimaryComboBox.ValueMember
                    = form.SpecifiedByConditionComboBox.ValueMember
                    = form.OtherNotSinglLabelComboBox.ValueMember ="Value";

                form.NoLabelBehaviorComboBox.SelectedValue = secondaryLabelCriteria.NoLabelBehavior;

                form.EqualsToPrimaryComboBox.SelectedValue = secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior;
                form.SpecifiedByConditionComboBox.SelectedValue = secondaryLabelCriteria.SpecifiedByConditionFileBehavior;
                form.OtherNotSinglLabelComboBox.SelectedValue = secondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior;

                if (form.ShowDialog() == DialogResult.OK)
                {
                    primaryLabelSpec.PrefixKey = form.PrimaryPrefixKeyTextBox.Text;
                    primaryLabelSpec.ItemNumberStartIndex = (int)Math.Round(form.PrimaryItemStartUpDown.Value);
                    primaryLabelSpec.ItemNumberLength = (int)Math.Round(form.PrimaryItemLengthUpDown.Value);
                    primaryLabelSpec.SerialNumberStartIndex = (int)Math.Round(form.PrimarySerialStartUpDown.Value);
                    primaryLabelSpec.SerialNumberLength = (int)Math.Round(form.PrimarySerialLengthUpDown.Value);

                    var acceptTypes = SecondaryLabelTypes.None;
                    acceptTypes |= form.AcceptC3LabelCheckBox.Checked ? SecondaryLabelTypes.C3Label : SecondaryLabelTypes.None;
                    acceptTypes |= form.AcceptSingleSymbolLabelCeckBox.Checked ? SecondaryLabelTypes.SingleSymbol : SecondaryLabelTypes.None;

                    secondaryLabelCriteria.AcceptableTypes = acceptTypes;

                    secondaryLabelCriteria.NoLabelBehavior = (SecondaryNoLabelBehavior)form.NoLabelBehaviorComboBox.SelectedValue;
                    secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior = (SecondaryLabelBehavior)form.EqualsToPrimaryComboBox.SelectedValue;
                    secondaryLabelCriteria.SpecifiedByConditionFileBehavior = (SecondaryLabelBehavior)form.SpecifiedByConditionComboBox.SelectedValue;
                    secondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior = (SecondaryLabelBehavior)form.OtherNotSinglLabelComboBox.SelectedValue;

                    Mode.ExtendedProperties[FixedLengthSpec.PropertyKeyForPrimary] = primaryLabelSpec;
                    Mode.ExtendedProperties[SecondaryLabelCriteria.PropertyKey] = secondaryLabelCriteria;
                }
            }
        }
    }
}

File: src/RepositoryModules.VB/ModeConfigForm1.vb
Imports Storex

Friend Class ModeConfigForm1
    Private Shared ReadOnly SecondaryLabelBehaviorViewModels As IReadOnlyList(Of KeyValuePair(Of String, SecondaryLabelBehavior)) = {
        New KeyValuePair(Of String, SecondaryLabelBehavior)("許可", SecondaryLabelBehavior.Default),
        New KeyValuePair(Of String, SecondaryLabelBehavior)("警告", SecondaryLabelBehavior.Warnning),
        New KeyValuePair(Of String, SecondaryLabelBehavior)("拒否", SecondaryLabelBehavior.Deny)
    }

    Sub New()
        InitializeComponent()
    End Sub

    Private Sub OkButton_Click(sender As Object, e As EventArgs)

        If String.IsNullOrEmpty(PrimaryPrefixKeyTextBox.Text.TrimEnd()) Then

            ErrorProvider.SetError(PrimaryPrefixKeyTextBox, "入力してください")
        Else

            ErrorProvider.SetError(PrimaryPrefixKeyTextBox, Nothing)
        End If

        DialogResult = DialogResult.OK
    End Sub

    Public Shared Sub Configure(Mode As IMode)

        Dim PrimaryLabelSpec As FixedLengthSpec = Nothing
        Dim SecondaryLabelCriteria As SecondaryLabelCriteria = Nothing

        PrimaryLabelSpec = If(Mode.TryExtractProperty(FixedLengthSpec.PropertyKeyForPrimary, PrimaryLabelSpec), PrimaryLabelSpec, New FixedLengthSpec())

        SecondaryLabelCriteria = If(Mode.TryExtractProperty(SecondaryLabelCriteria.PropertyKey, SecondaryLabelCriteria), SecondaryLabelCriteria, New SecondaryLabelCriteria())

        Using form = New ModeConfigForm1()

            form.PrimaryPrefixKeyTextBox.Text = PrimaryLabelSpec.PrefixKey
            form.PrimaryItemStartUpDown.Value = PrimaryLabelSpec.ItemNumberStartIndex
            form.PrimaryItemLengthUpDown.Value = PrimaryLabelSpec.ItemNumberLength
            form.PrimarySerialStartUpDown.Value = PrimaryLabelSpec.SerialNumberStartIndex
            form.PrimarySerialLengthUpDown.Value = PrimaryLabelSpec.SerialNumberLength

            form.AcceptC3LabelCheckBox.Checked = (SecondaryLabelCriteria.AcceptableTypes And SecondaryLabelTypes.C3Label) = SecondaryLabelTypes.C3Label
            form.AcceptSingleSymbolLabelCeckBox.Checked = (SecondaryLabelCriteria.AcceptableTypes And SecondaryLabelTypes.SingleSymbol) = SecondaryLabelTypes.SingleSymbol

            form.EqualsToPrimaryComboBox.DataSource = SecondaryLabelBehaviorViewModels.ToList()
            form.SpecifiedByConditionComboBox.DataSource = SecondaryLabelBehaviorViewModels.ToList()
            form.OtherNotSinglLabelComboBox.DataSource = SecondaryLabelBehaviorViewModels.ToList()

            form.NoLabelBehaviorComboBox.DataSource = {
                New KeyValuePair(Of String, SecondaryNoLabelBehavior)("許可", SecondaryNoLabelBehavior.Default),
                New KeyValuePair(Of String, SecondaryNoLabelBehavior)("警告", SecondaryNoLabelBehavior.Warnning),
                New KeyValuePair(Of String, SecondaryNoLabelBehavior)("拒否", SecondaryNoLabelBehavior.Deny),
                New KeyValuePair(Of String, SecondaryNoLabelBehavior)("タグで品番照合 (合致しない場合警告) ", SecondaryNoLabelBehavior.WarningWhenTagNotMatched),
                New KeyValuePair(Of String, SecondaryNoLabelBehavior)("タグで品番照合 (合致しない場合拒否)", SecondaryNoLabelBehavior.DenyWhenTagNotMatched)
            }

            form.NoLabelBehaviorComboBox.DisplayMember = NameOf(KeyValuePair(Of String, SecondaryNoLabelBehavior).Key)
            form.EqualsToPrimaryComboBox.DisplayMember = NameOf(KeyValuePair(Of String, SecondaryLabelBehavior).Key)
            form.SpecifiedByConditionComboBox.DisplayMember = NameOf(KeyValuePair(Of String, SecondaryLabelBehavior).Key)
            form.OtherNotSinglLabelComboBox.DisplayMember = NameOf(KeyValuePair(Of String, SecondaryLabelBehavior).Key)

            form.NoLabelBehaviorComboBox.ValueMember = NameOf(KeyValuePair(Of String, SecondaryNoLabelBehavior).Value)
            form.EqualsToPrimaryComboBox.ValueMember = NameOf(KeyValuePair(Of String, SecondaryLabelBehavior).Value)
            form.SpecifiedByConditionComboBox.ValueMember = NameOf(KeyValuePair(Of String, SecondaryLabelBehavior).Value)
            form.OtherNotSinglLabelComboBox.ValueMember = NameOf(KeyValuePair(Of String, SecondaryLabelBehavior).Value)

            form.NoLabelBehaviorComboBox.SelectedValue = SecondaryLabelCriteria.NoLabelBehavior

            form.EqualsToPrimaryComboBox.SelectedValue = SecondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior
            form.SpecifiedByConditionComboBox.SelectedValue = SecondaryLabelCriteria.SpecifiedByConditionFileBehavior
            form.OtherNotSinglLabelComboBox.SelectedValue = SecondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior

            If form.ShowDialog() = DialogResult.OK Then

                PrimaryLabelSpec.PrefixKey = form.PrimaryPrefixKeyTextBox.Text
                PrimaryLabelSpec.ItemNumberStartIndex = Math.Round(form.PrimaryItemStartUpDown.Value)
                PrimaryLabelSpec.ItemNumberLength = Math.Round(form.PrimaryItemLengthUpDown.Value)
                PrimaryLabelSpec.SerialNumberStartIndex = Math.Round(form.PrimarySerialStartUpDown.Value)
                PrimaryLabelSpec.SerialNumberLength = Math.Round(form.PrimarySerialLengthUpDown.Value)

                Dim AcceptTypes = SecondaryLabelTypes.None
                AcceptTypes = AcceptTypes Or If(form.AcceptC3LabelCheckBox.Checked, SecondaryLabelTypes.C3Label, SecondaryLabelTypes.None)
                AcceptTypes = AcceptTypes Or If(form.AcceptSingleSymbolLabelCeckBox.Checked, SecondaryLabelTypes.SingleSymbol, SecondaryLabelTypes.None)

                SecondaryLabelCriteria.AcceptableTypes = AcceptTypes

                SecondaryLabelCriteria.NoLabelBehavior = CType(form.NoLabelBehaviorComboBox.SelectedValue, SecondaryNoLabelBehavior)

                SecondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior = CType(form.EqualsToPrimaryComboBox.SelectedValue, SecondaryLabelBehavior)
                SecondaryLabelCriteria.SpecifiedByConditionFileBehavior = CType(form.SpecifiedByConditionComboBox.SelectedValue, SecondaryLabelBehavior)
                SecondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior = CType(form.OtherNotSinglLabelComboBox.SelectedValue, SecondaryLabelBehavior)

                Mode.ExtendedProperties(FixedLengthSpec.PropertyKeyForPrimary) = PrimaryLabelSpec
                Mode.ExtendedProperties(SecondaryLabelCriteria.PropertyKey) = SecondaryLabelCriteria
            End If
        End Using
    End Sub
End Class
